### PR TITLE
[SPARK-7958] [STREAMING] Handled exception in StreamingContext.start() to prevent leaking of actors

### PR DESCRIPTION
**Pull request by tdas**
StreamingContext.start() can throw exception because DStream.validateAtStart() fails (say, checkpoint directory not set for StateDStream). But by then JobScheduler, JobGenerator, and ReceiverTracker has already started, along with their actors. But those cannot be shutdown because the only way to do that is call StreamingContext.stop() which cannot be called as the context has not been marked as ACTIVE.

The solution in this PR is to stop the internal scheduler if start throw exception, and mark the context as STOPPED.

### CODE REVIEW

**Comment by tdas** [comment]
@zsxwing Please take a look.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 1]
```diff
@@ -126,6 +126,8 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
     eventLoop.post(ErrorReported(msg, e))
   }
 
+  def isStarted(): Boolean = eventLoop != null
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: It's better to add `synchronized` here. Although the test only uses one thread now, someone may use it in multiple threads in future.

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: Good catch. Added.

**Comment by zsxwing** [comment]
LGTM

**Comment by tdas** [comment]
Thanks @zsxwing . Merging this to both master and branch-1.4

**Pull request closed without merging**